### PR TITLE
unit test for bumping nonces - issue #10

### PR DESCRIPTION
**Pull request by pdobacz**
This is a unit test for EthermintApplication which is supposed to make sure that #10 never comes back (it's been fixed in `major-update`). The issue was about nonces not having been handled correctly, when transactions were coming in from two "input flows", namely the ABCI and geth's RPC.

It required a minor modification of `ethereum.Backend` setup to handle the RPC client as an injected dependency.

This is the cleanest state of the test we were able to get to with our current golang- and ethermint-lore. Comments and guidance much appreciated :), especially in removing the configuration cruft in setup.

### CODE REVIEW

**Comment by illya13** [comment]
can you squash commit ?

**Comment by pdobacz** [comment]
done. Do you maybe think it'd be a good idea to include a `go test` step in the travis.yml right away in this PR already?

File: ethereum/backend.go [hunk 1]
```diff
@@ -19,13 +19,18 @@ import (
...
 	abciTypes "github.com/tendermint/abci/types"
-	client "github.com/tendermint/go-rpc/client"
 	core_types "github.com/tendermint/tendermint/rpc/core/types"
 	emtTypes "github.com/tendermint/ethermint/types"
 	"github.com/ethereum/go-ethereum/core/vm"
 	"github.com/ethereum/go-ethereum/params"
 )
 
+// used by Backend to call tendermint rpc endpoints
+type Client interface {
```
Comment (melekes) [inline ethereum/backend.go:29]: precisely what I was talking about here https://github.com/tendermint/go-rpc/issues/8 :(
let's keep it for now though.

@pdobacz could you add the `TODO: replace with HttpClient https://github.com/tendermint/go-rpc/issues/8`

File: app/app_test.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	switch method {
+	case "status":
+		*tmresult = &core_types.ResultStatus{}
+		return tmresult, nil
+	case "broadcast_tx_sync":
+		mockclient.Broadcast_tx_sync_called = true
+		*tmresult = &core_types.ResultBroadcastTx{}
+		return tmresult, nil
+	}
+	return tmresult, abciTypes.ErrInternalError
```
Comment (melekes) [inline app/app_test.go:49]: why not `nil, abciTypes.ErrInternalError`?

File: app/app_test.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	assert.Equal(t, app.CheckTx(encodedtx), abciTypes.OK)
+	app.BeginBlock([]byte{}, &abciTypes.Header{Height: height})
+	assert.Equal(t, app.DeliverTx(encodedtx), abciTypes.OK)
+	app.EndBlock(height)
+	app.Commit()
+
+	// replays should fail - we're checking if the transaction got through earlier, by replaying the nonce
+	assert.Equal(t, app.CheckTx(encodedtx), abciTypes.ErrInternalError)
+	// ...on both interfaces of the app
+	assert.Equal(t, backend.Ethereum().ApiBackend.SendTx(ctx, tx1), errors.New("Nonce too low"))
```
Comment (ebuchman) [inline app/app_test.go:94]: Is there a better way to represent this error than the hardcoded string?

**Pull request merged**